torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -34,47 +35,66 @@ int64_t wrapDim(int64_t dim, at::IntList sizes) {
   return dim;
 }
 
-at::Tensor representativeTensor(const TensorType * type) {
-  auto backend = type->device() == -1 ? at::kCPU : at::kCUDA;
-  at::DeviceGuard device_guard(type->device());
-  auto & attype = at::getType(backend, type->scalarType());
-  return attype.tensor(type->sizes(), type->strides()).zero_();
+IValue representativeValue(Value* v) {
+  TypePtr type_ = v->type();
+  // if the value is actually constant, just use it!
+  if(auto iv = toIValue(v)) {
+    return *iv;
+  }
+  if (TensorType* type = type_->cast<TensorType>()) {
+    auto backend = type->device() == -1 ? at::kCPU : at::kCUDA;
+    at::DeviceGuard device_guard(type->device());
+    auto& attype = at::getType(backend, type->scalarType());
+    return attype.tensor(type->sizes(), type->strides()).zero_();
+  } else if (type_->isSubtypeOf(*FloatType::get())) {
+    return 0.f;
+  }
+  // we should not get here because isValidArgumentForRunning should have
+  // prevented it
+  std::stringstream ss;
+  ss << "unable to create representative value for: " << type_->str()
+     << ". File a bug report.";
+  throw std::runtime_error(ss.str());
 }
 
 void PropagateShapeOnBlock(Block * block, bool insert_expands=true);
 
+// for each node in the schema with type Tensor, extract the TensorType
+// returns at::nullopt if any Tensor in the schema does not have a known shape
+// ignores non-tensor in the list of inputs
 at::optional<std::vector<TensorType*>> gatherTensorTypes(Node *node) {
   std::vector<TensorType*> tensor_types;
-  tensor_types.reserve(node->inputs().size());
-  // TODO (apaszke): Remove once we stop using attributes
-  // XXX: we also make the exception for cat, because we need shape prop to work for it
-  // (we have tests). We'll have to remove the special case once we stop flattening lists into inputs.
-  if (node->hasAttributes() || node->kind() == aten::cat) {
-    std::vector<Value*> inputs = node->inputs();
-    if (node->kind() == aten::cat && inputs.back()->type()->isSubtypeOf(*IntType::get())) {
-      inputs.pop_back();
-    }
-    for (Value *v : inputs) {
-      TensorType* type = v->type()->cast<TensorType>();
-      if(!type) return at::nullopt;
-      tensor_types.push_back(type);
-    }
-  } else {
-    auto & schema = node->schema();
-    auto & args = schema.arguments;
-    // XXX: This gets triggered for nodes that have Tensor[] as arguments.
-    // Those are currently very annoying to handle, because the lists are simply
-    // inlined into the node inputs, so we bail out from shape propagation for now.
-    if (schema.is_vararg || args.size() != node->inputs().size()) {
-      return at::nullopt;
+
+  auto & schema = node->schema();
+  auto & args = schema.arguments;
+  // can't handle varargs primitives because we don't know what should be a Tensor
+  if (schema.is_vararg) {
```

Comment:
We currently use `vararg` schemas only for `prim::` ops, right?

Reply:
Yes,  only primitives and it will likely stay that way. It is just easier to do it this way than to have a whole if statement at the beginning of any schema check. It is easier to define varargs functions as those which have zero type-checked arguments and unlimited unchecked arguments.